### Investigating the discharge summary

In [3]:
import pandas as pd
import numpy as np
import os
import psycopg2
import sqlalchemy
from sqlalchemy import create_engine
import string

In [4]:
# connect to the mimic database and set the search path to the 'mimiciii' schema

dbschema='mimiciii'
cnx = create_engine('postgresql+psycopg2://aa5118:mimic@localhost:5432/mimic',
                    connect_args={'options': '-csearch_path={}'.format(dbschema)})


In [5]:
cat = "'Discharge summary'"
df_temp = pd.read_sql_query('''
  SELECT hadm_id FROM noteevents WHERE category = ''' + cat + '''
''', cnx)
df_temp.shape

(59652, 1)

59652 Discharge summaries from ~2m notes

In [6]:
df_temp = pd.read_sql_query('''
  SELECT hadm_id FROM noteevents WHERE category = ''' + cat + ''' AND hadm_id IS NOT NULL
  ORDER BY hadm_id
''', cnx)
df_temp.shape

(59652, 1)

These all contain a hospital admission ID (`hadm_id`) - there are no NULLs. Only patients who get admitted to the ICU during their hospital admission will have an hadm_id in this table according to https://github.com/MIT-LCP/mimic-code/issues/237. So this means the discharge summary is only for being discharged from the ICU, not the hospital.

In [8]:
df_temp = pd.read_sql_query('''
  SELECT COUNT(DISTINCT hadm_id) FROM noteevents WHERE category = ''' + cat + ''' AND hadm_id IS NOT NULL
  --ORDER BY hadm_id
''', cnx)
df_temp

count
0  52726

However, there are only 52726 unique `hadm_id`s in the table. This shows that patients can have multiple distinct ICU stays during a single hospital admission. Indeed there are ~7k duplicate `hadm_id`s. And for each ICU stay, they will get a discharge summary. This clears the picture a litte with deciding how we can utilise the predictive power of the discharge summaries

In [11]:
# breakdown of note categories showing the number of notes, average number of characters
# and the number of notes in each category where the note time was provided

df_summary = pd.read_sql_query('''
  SELECT 
      category,
      COUNT(category),
      ROUND(AVG(LENGTH(text)),-1)::integer AS text_avg_chars,
      ROUND(AVG(array_length(regexp_split_to_array(text, '\s+'), 1)))::integer AS text_avg_words,
      COUNT(charttime) AS time_provided
  FROM noteevents
  GROUP BY category
''', cnx)
df_summary

category   count  text_avg_chars  text_avg_words  time_provided
0    Case Management      967            1120             162            967
1             Consult      98            6040             761             98
2   Discharge summary   59652            9620            1436              0
3                 ECG  209051             210              30              0
4                Echo   45794            2320             326              0
5             General    8301            1560             216           8260
6             Nursing  223556            1790             264         222172
7       Nursing/other  822497             800             132         822497
8           Nutrition    9418            2430             321           9411
9            Pharmacy     103            2580             357            102
10         Physician   141624            7140             858         141048
11          Radiology  522279            1740             209         522279
12     Rehab Services    5431            3120             436           5429
13       Respiratory    31739            1360             155          31703
14        Social Work    2670            2160             333           2648

In [14]:
print(df_summary.sum(axis = 0, skipna = True))
print ("\n")
print(df_summary.mean(axis = 0, skipna = True))

category          Case Management ConsultDischarge summaryECGEch...
count                                                       2083180
text_avg_chars                                                43990
text_avg_words                                                 5996
time_provided                                               1766614
dtype: object


count             138878.666667
text_avg_chars      2932.666667
text_avg_words       399.733333
time_provided     117774.266667
dtype: float64


In [18]:
print("Char wgtd avg count:", sum(df_summary['text_avg_chars'] * df_summary['count'])/sum(df_summary['count']))
print("Word wgtd avg count:", sum(df_summary['text_avg_words'] * df_summary['count'])/sum(df_summary['count']))

Char wgtd avg count: 1826.9048521971217
Word wgtd avg count: 248.84047705911155


Discharge summaries are the longest note category in the `NOTEEVENTS` table with almost 10k characters.

In [23]:
# confirming that the dataframe output should have 0 rows

df_temp = pd.read_sql_query('''
  SELECT charttime FROM noteevents WHERE category = ''' + cat + ''' AND charttime IS NOT NULL
''', cnx)
df_temp

Empty DataFrame
Columns: [charttime]
Index: []

Furthermore, they are one of the few note categories which do not provide a timestamp

#### Cause of death

Below query matches the string "cause of death" in the text field and returns the substring occuring after it until the end of the note

In [55]:
sql = """
SELECT 
    SUBSTRING (n.text FROM 
        POSITION('cause of death' IN n.text) FOR (LENGTH(n.text) - POSITION('cause of death' IN n.text)))
FROM noteevents n
INNER JOIN admissions a
ON n.hadm_id = a.hadm_id
WHERE a.hospital_expire_flag = 1
AND lower(n.text) LIKE '%cause of death%'
AND category LIKE 'Discharge summary%';
"""

df_temp = pd.read_sql_query(sqlalchemy.text(sql), cnx)
df_temp.head()

substring
0  cause of death was hyperkalemia from acute ren...
1  cause of death was likely cardiac arrest.\nSec...
2  cause of death was cardiac arrest secondary to...
3  cause of death being cardiopulmonary\nfailure ...
4  Admission Date:  [**2131-9-2**]              D...

In [56]:
print([word for word in df_temp.iloc[1]])

['cause of death was likely cardiac arrest.\nSecondary cause sepsis.\n\n\n\n\n                          [**Name6 (MD) **] [**Name8 (MD) **], M.D.  [**MD Number(1) 968**]\n\nDictated By:[**Last Name (NamePattern1) 2584**]\n\nMEDQUIST36\n\nD:  [**2140-6-9**]  04:23\nT:  [**2140-6-12**]  16:16\nJOB#:  [**Job Number 2585**]']


Maybe we can use the final discharge summary from an ICU visit to predict whether the patient will die in the ICU at their next visit.

In [22]:
# breakdown by gender

sql = """
  SELECT p.gender, 
         COUNT(p.gender), 
         AVG(ROUND((cast(n.chartdate as date) - cast(p.dob as date)) / 365.242,0)) AS age,
         COUNT(DISTINCT(n.subject_id)) AS unique_subjects
  FROM patients p 
  INNER JOIN noteevents n 
  ON p.subject_id = n.subject_id
  WHERE n.category = 'Discharge summary'
  AND ROUND((cast(chartdate as date) - cast(dob as date)) / 365.242,0) > 14
  GROUP BY p.gender
"""

df_summary = pd.read_sql_query(sqlalchemy.text(sql), cnx)

df_summary

gender  count        age  unique_subjects
0      F  24291  81.510395            16238
1      M  31113  70.147784            21162

There seems to be a reasonably large difference in the number of men and women as well as their average ages. So we should stratify when we split our dataset to ensure that the training, validation and test sets all have approximately the same average age and gender balance.